src/Microsoft.AspNet.Hosting/WebApplicationBuilder.cs

```diff
@@ -246,5 +185,78 @@ public WebHostBuilder UseStartup(Action<IApplicationBuilder> configureApp, Actio
                 });
             return this;
         }
+
+        public IWebApplication Build()
+        {
+            var hostingServices = BuildHostingServices();
+
+            var hostingContainer = hostingServices.BuildServiceProvider();
+
+            var appEnvironment = hostingContainer.GetRequiredService<IApplicationEnvironment>();
+            var startupLoader = hostingContainer.GetRequiredService<IStartupLoader>();
+
+            _options = new WebHostOptions(_config);
+
+            // Initialize the hosting environment
+            _options.WebRoot = _webRoot ?? _options.WebRoot;
+            _hostingEnvironment.Initialize(appEnvironment.ApplicationBasePath, _options, _config);
+
+            if (!string.IsNullOrEmpty(_environmentName))
+            {
+                _hostingEnvironment.EnvironmentName = _environmentName;
+            }
+
+            var application = new WebApplication(hostingServices, startupLoader, _options, _config);
+
+            // Only one of these should be set, but they are used in priority
+            application.Server = _server;
+            application.ServerFactory = _serverFactory;
+            application.ServerFactoryLocation = _options.Server ?? _serverFactoryLocation;
```

Comment:
You already passed _options into WebApplication, it's weird to also use it to set properties.